NYTimesMostViewedArticles/NetworkClient.swift

```diff
@@ -0,0 +1,78 @@
+//
+//  NetworkClient.swift
+//  NYTimesMostViewedArticles
+//
+//  Created by Sagar Pahwa on 17/11/20.
+//  Copyright © 2020 Sagar Pahwa. All rights reserved.
+//
+
+import Foundation
+
+protocol NetworkClientType {
+    func getMostViewedArticles(period: PeriodSection, completion: @escaping (Response?, Error?)->())
+    
+    func url(period: PeriodSection) -> URL
+}
+
+enum Keys: String {
+    case nyTimesKey = "yxPlqAEPWW3qKi0BgofuzzXeEbxkGfj4"
+    case mockKey = "1234567890"
+}
+
+enum PeriodSection: Int {
+    case day = 1
+    case week = 7
+    case month = 30
+}
+
+struct API {
+    static let baseURL = "http://api.nytimes.com/"
+    static let contentURL = "svc/mostpopular/v2/viewed"
+        //"svc/mostpopular/v2/mostviewed/all-sections/"
+}
+
+
+typealias closure = ()
+
+extension NetworkClientType {
+    func url(period: PeriodSection) -> URL {
+        let string = "\(API.baseURL)/\(API.contentURL)/\(period.rawValue).json?api-key=\(Keys.nyTimesKey.rawValue)"
+        return URL(string: string)!
+    }
+}
+
+class NetworkClient: NetworkClientType {
+    
+    let session = URLSession.shared
+    
+
+    func getMostViewedArticles(period: PeriodSection, completion: @escaping (Response?, Error?)->()) {
+
+        var request = URLRequest(url: self.url(period: period))
+        request.httpMethod = "GET"
```

Comment:
1. All this general response parsing i.e. checking response status, mime, etc should not be part of each API method. Rather a root level network client should be written which will take care of all such things and will return json or error which respective network client functions will cast into required data model.

2. Also please try to use `Result` generic enum as defined here https://developer.apple.com/documentation/swift/result

Reply:
2nd point is still unresolved.